### DISCUSSION TITLE
Unable to complete "make" command on FreeBSD 13.1; libsharpyuv.so: undefined reference to floorf

### DISCUSSION THREAD

**Discussion by aredon** [Motion-Project/motion#1641] [Q&A] [answered]
### Did you read the guide?

Yes

### What is the base version number of Motion being used?

4.5.x

### What was the install method?

Built from source code

### What is base architecture?

x86

### What is the distro being used?

Other

### Disto version number

FREEBSD-13.1

### Camera type(s) being used?

Network camera with RTSP

### Describe the issue/problem and steps to reproduce

Following the instructions in the guide for building on freeBSD systems there appears to be some missing functions in a library file. The file does exist on my system but it would seem that floorf is missing. 

I issued the make command and this happened:
![image](https://user-images.githubusercontent.com/1328757/218940945-c067dccd-1f67-4eaf-bd3e-cd8fd61ee087.png)

I was previously running motion on FreeBSD-11.3 without issue but it stopped working after a system upgrade. I have tried make clean and I have tried installing via pip (which for some reason installs motion v0.2.0 which is wild). 

### Relevant Motion log output (at log_level 8)

```shell
wscale -lavdevice  -L/usr/local/lib/mysql -lmysqlclient  -L/usr/local/lib -lpq-L/usr/local/lib -lsqlite3
ld: error: /usr/local/lib/libsharpyuv.so: undefined reference to floorf [--no-allow-shlib-undefined]
ld: error: /usr/local/lib/libsharpyuv.so: undefined reference to pow [--no-allow-shlib-undefined]
cc: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error code 1

Stop.
make[2]: stopped in /root/motion/src
*** Error code 1

Stop.
make[1]: stopped in /root/motion
*** Error code 1

Stop.
make: stopped in /root/motion
```

**Comment by Mr-Dave** [comment]
FYI.  Moved to a discussion since I just tried multiple configuration options on FreeBSD 13.1 and they all compiled with issue. Seems like something unique to the setup.

**Reply by aredon** [reply]
Hmmm. This is freebsd running in an iocage jail but I'm not sure why that would matter. Send me your library file?

**Comment by Mr-Dave** [comment]
Well, I'm not sure I'd be able to send you the library file.  I think that it is important to note that Motion does not use that library directly. It would be used only indirectly based upon libs that Motion does use.  Reviewing the file name, I'd guess that it is used by libjpeg, ffmpeg or webp so you may want to check the integrity of the installs for those.

One other note:

> I have tried installing via pip (which for some reason installs motion v0.2.0 which is wild).

pip is a python application installer.  Motion is not a python app and the v0.2.0 is some python application with the same name that has nothing to do with this Motion application.  All the installs here would be using `pkg install`

**Reply by aredon** [reply]
When last I looked I came up empty for which package that library file even comes from. I have tried pkg install for each of the motion prerequisites individually and it tells me they are all up to date. I'll try creating a fresh 13.1 jail and see if I can install motion there from scratch - that'll basically prove out your theory.

**Reply by aredon** [reply]
Ok. I started a completely fresh iocage jail of FreeBSD-13.1.
<img width="691" alt="image" src="https://user-images.githubusercontent.com/1328757/219459792-136797ac-24e9-4170-b003-a1cfb9f7cbb0.png">

The very first commands I ran were these sans sudo because the jail is root. Additionally pkg had to bootstrap itself to install because it was a fresh system:
<img width="520" alt="image" src="https://user-images.githubusercontent.com/1328757/219458312-5732d00d-a791-4925-a5b6-c3370523c300.png">

After getting to the Make step and hitting enter I encountered the exact same issue:
<img width="269" alt="image" src="https://user-images.githubusercontent.com/1328757/219458402-fc02bb05-5994-4367-b97c-44d5947db05b.png">

Is there any version difference between what you ran and what I ran? I can't think of any reason why the fresh iocage jail should differ at all from what you did?

**Comment by Mr-Dave** [comment]
The log of the compile is [here](https://gist.github.com/Mr-Dave/c5da8df61659612459d0d98de9c00a10).  I do install xfce, slim and firefox in my FreeBSD test machine

**Reply by aredon** [reply]
Well iocage jails are headless but I tried installing xfce, slim, and firefox anyway and it didn't make a difference. I've created a brand new jail here in order to give you a full clean log. About the only thing I noticed that's different is that you're calling a --prefix on your ./configure but the jail only has a root user and thus /usr/local should be the correct prefix for those libraries anyway. 

https://gist.github.com/aredon/a60072f90fc1c5cfbd16b18b90605842

**Comment by Mr-Dave** [comment, marked as answer]
The good news is that I was able to reproduce this by not installing xfce.  The offending package is webp.  Specifically, it seems that the raw install of webp puts in a pkgconf file that indicates a requirement for libsharpyuv.  In my FreeBSD install that first starts with installing xfce, that statement is not there.

For reference, the package config file is what the configure uses in order to determine dependencies and libs.  The one for webp is located at /usr/local/libdata/pkgconfig/libwebp.pc

As for options, you can remove support for webp from Motion by using `./configure --without-webp`  Alternatively, you could edit the libwebp.pc file and remove the requires line(I do not know if there are consequences of this or not).  


EDIT:
Looking at the github source for the libwebp.pc file [here](https://github.com/webmproject/libwebp/blob/1.2.4/src/libwebp.pc.in) it seems to be that the requires statement isn't needed for the version installed with FreeBSD.

**Reply by aredon** [reply]
Could you walk me through how you worked out it was webp? I searched for ages for which package was related to that library.

I admit I must have flubbed the xfce install. It took a while so I walked away, came back later and might have closed the console it was running in, but later when I checked it said the package was installed so idk.

I'll try again tonight without webp in the motion compile. Seems pretty non-critical for my application anyway.

**Comment by Mr-Dave** [comment]
It was the sequence that it was listed.    After the webp libs and before the ffmpeg.  After that I used the command 
``` pkgconf --libs libwebp``` to see which was adding the -lsharpyuv

```
CFLAGS: -D_THREAD_SAFE -I/usr/local/include  -I/usr/local/include  -I/usr/local/include -I/usr/local/include/webp  -I/usr/local/include  -g -O2

LIBS:  -pthread   -L/usr/local/lib -ljpeg  -L/usr/local/lib -lmicrohttpd  -L/usr/local/lib -lwebpmux -lwebp -lsharpyuv  -L/usr/local/lib -lavutil -lavformat -lavcodec -lswscale -lavdevice

LDFLAGS:
```

I'd note that I've subsequently learned that it is not xfce.  It has to do with the version 1.2.4 that I installed on my system in November did not have this requirement.  But now the new version 1.2.4 does have this.  A bug has been submitted to FreeBSD but the initial response has been "probably need to refresh your tree".

**Reply by aredon** [reply]
Yeah no iocage pulls a fresh tree when the jail is created so it's definitely not that. 😅 That's a clever thought that I wouldn't have considered so I appreciate the insight. 

I've successfully built it without webp now! Thanks for your help.
